src/selectors/ContainerSelector.h

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file ContainerSelector.h
+ * @author F. Gratl
+ * @date 11.06.18
+ */
+
+#pragma once
+
+#include <containers/DirectSum.h>
+#include <containers/LinkedCells.h>
+#include <containers/ParticleContainer.h>
+#include <containers/VerletLists.h>
+#include <array>
+#include <vector>
+namespace autopas {
+
+/**
+ * Possible choices for the particle container type.
+ */
+enum ContainerOptions {
+  directSum = 0,
+  linkedCells = 1,
+  verletLists = 2,
+};
+
+/**
+ * Provides a way to iterate over the possible choices of ContainerOption.
+ */
+static std::vector<ContainerOptions> allContainerOptions = {ContainerOptions::directSum, ContainerOptions::linkedCells,
+                                                            ContainerOptions::verletLists};
+
+/**
+ * Selector for a particle container.
+ *
+ * The class is given a list of allowed container and traversal options to choose from.
+ * This class selects the optimal container and delegates the choice of the optimal traversal down to this container.
+ *
+ * @tparam Particle
+ * @tparam ParticleCell
+ */
+template <class Particle, class ParticleCell>
+class ContainerSelector {
+ public:
+  /**
+   * Constructor for the ContainerSelecor class.
+   * @param boxMin Lower corner of the container.
+   * @param boxMax Upper corner of the container.
+   * @param cutoff  Cutoff radius to be used in this container.
+   * @param verletSkin Length added to the cutoff for the verlet lists' skin.
+   * @param verletRebuildFrequency Specifies after how many pair-wise traversals the neighbor lists are to be rebuild.
+   * @param allowedContainerOptions Vector of container types the selector can choose from.
+   * @param allowedTraversalOptions Vector of traversals the selector can choose from.
+   */
+  ContainerSelector(std::array<double, 3> &boxMin, std::array<double, 3> &boxMax, double cutoff, double verletSkin,
+                    unsigned int verletRebuildFrequency, std::vector<ContainerOptions> allowedContainerOptions,
+                    std::vector<TraversalOptions> allowedTraversalOptions)
+      : _boxMin(boxMin),
+        _boxMax(boxMax),
+        _cutoff(cutoff),
+        _verletSkin(verletSkin),
+        _verletRebuildFrequency(verletRebuildFrequency),
+        _allowedContainerOptions(std::move(allowedContainerOptions)),
+        _allowedTraversalOptions(std::move(allowedTraversalOptions)),
+        _optimalContainer(nullptr) {}
+
+  /**
+   * Getter for the optimal container.
+   * @return Smartpointer to the optimal container.
+   */
+  std::shared_ptr<ParticleContainer<Particle, ParticleCell>> getOptimalContainer();
+
+  /**
+   * Evaluates the optimal container option.
```

Comment:
is this function actually allowed to modify the particles? if it does, there has to be some kind of way to rezero the resulting forces, etc.
-> add new issue

Reply:
needs to be thought about for the next pull request.

Reply:
#71